x-pack/legacy/plugins/apm/server/lib/helpers/es_client.ts

```diff
@@ -66,8 +66,12 @@ async function getParamsForSearchRequest(
   apmOptions?: APMOptions
 ) {
   const uiSettings = req.getUiSettingsService();
+  const { server } = req;
   const [indices, includeFrozen] = await Promise.all([
-    getApmIndices(req.server),
+    getApmIndices({
+      config: server.config(),
+      savedObjectsClient: server.savedObjects.getScopedSavedObjectsClient(req)
+    }),
```

Comment:
What's the reason for changing the signature of `getApmIndices`? What about keeping `getApmIndices(server)`?

Reply:
Ah, I guess this is because you are sometimes injecting a scoped saved objects client, and other times you want to use the unscoped (internalUser) one? 

Is currently any cases where we want `getApmIndices` to use the unscoped saved objects client? If not I think you should keep the saved objects logic inside `getApmIndices`.

Reply:
For NP migration readiness we recommend making your dependencies on `server` explicit, so instead of injecting `server` deep into your plugin, injecting `{config, savedObjectsClient}` deep into your plugin will make it easier to see which dependencies you're actually using.

Reply:
Doesn't this go again the suggestion to ["preserve whole object"](https://refactoring.guru/preserve-whole-object)?

I don't follow all design patterns blindly but in this case I don't see the reasoning for repeatedly picking the same properties off `server` and then passing them along. Since `server` is a typed object we can have good confidence in the properties being available "downstream". If we worry about the object being mutated we can use the `Readonly` typescript generic.

Reply:
the suggestion is to keep using a single object, but to construct that object yourself by picking from `server`. So something like `const apmDependencies = {config: server.config()}`. You can do the same with typescript, as long as you create a new type which picks from `Legacy.Server` instead of the whole type as is.

Even though it follows our pattern, on it's own this change isn't so useful when only applied to a single function, so it might be best to create a type / variable for all APM's dependencies in a separate NP migration PR.

Reply:
I think the important part of the refactoring technique is:

> After you apply this refactoring technique, the code for getting all necessary data will be stored in one place – the method itself.

What I'd like to avoid is doing this 10 times:

```js
const apmDependencies = {
 config: server.config(),
 savedObjectsClient: server.savedObjects.getScopedSavedObjectsClient(req)
};
getApmIndices(apmDependencies);
```

and instead doing this 10 times:
```js
getApmIndices(server);
```

Reply:
>@sqren: Is currently any cases where we want getApmIndices to use the unscoped saved objects client? If not I think you should keep the saved objects logic inside getApmIndices.

We use the unscoped SOC when persisting the telemetry object.

Reply:
> We use the unscoped SOC when persisting the telemetry object.

But is that related to `getApmIndices`?

Reply:
`createApmAgentConfigurationIndex` on plugin init is the only place where we pass the unscoped, internal SavedObjectsClient to `getApmIndices` since there is no request context, only the server context. This is the primary reason i broke it down to `{ savedObjectsClient, config }` as the param for that function. I did pass earlier where `getApmIndices` only required a param of `req` but that would only work on API endpoints. I don't mind making 2 functions that do this for the sake of simple params.

Reply:
Okay I must have missed that. Then let's just leave it like it is.